lib/Sema/CSDiag.cpp

```diff
@@ -7200,18 +7200,19 @@ bool FailureDiagnosis::visitBindOptionalExpr(BindOptionalExpr *BOE) {
 }
 
 bool FailureDiagnosis::visitIfExpr(IfExpr *IE) {
-  auto typeCheckClauseExpr = [&](Expr *clause) -> Expr * {
-    return typeCheckChildIndependently(clause, Type(), CTP_Unused, TCCOptions(),
+  auto typeCheckClauseExpr = [&](Expr *clause, ContextualTypePurpose convertTypePurpose) -> Expr * {
```

Comment:
You should type check the condition as unused then and just use the purpose for the branches.